compiler/rustc_typeck/src/check/callee.rs

```diff
@@ -350,6 +350,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     }
                 }
 
+                let callee_ty = self.resolve_vars_if_possible(callee_ty);
```

Comment:
I wonder why callee_ty is an TyVar and not an IntVar, I think we should dig into this a bit instead of just patching up the diagnostic

Reply:
I'm guessing the `TyVar` is coming from the `{}` block's tail expression not having unified yet?

Reply:
If that were the case, wouldn't we also see the tyvar, when using a function type?

I think we're doing something odd when type checking the literal expression.

I'm currently somewhere deep in that code anyway, I'll check and post my findings

Reply:
Yea, you're right. The tail block expression will fall back to a type variable in case we don't know anything about the type. This happens in https://github.com/rust-lang/rust/blob/3982ac22497efda2bddf261021e88aa37b55e47b/compiler/rustc_typeck/src/check/expectation.rs#L119

So... I guess we can't avoid this at creation time. oh well...